File: package-lambda.js
#!/usr/bin/env node

/**
 * Package application for AWS Lambda deployment
 */

import { execSync } from 'child_process';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

console.log('\n📦 Packaging for AWS Lambda...\n');

try {
  // Clean up previous package
  console.log('🧹 Cleaning up previous package...');
  if (fs.existsSync('lambda-package.zip')) {
    fs.unlinkSync('lambda-package.zip');
  }
  if (fs.existsSync('lambda-temp')) {
    fs.rmSync('lambda-temp', { recursive: true, force: true });
  }

  // Create temp directory
  console.log('📁 Creating package directory...');
  fs.mkdirSync('lambda-temp');

  // Copy source files
  console.log('📄 Copying source files...');
  const filesToCopy = [
    'lambda-handler.js',
    'src',
    'package.json'
  ];

  for (const file of filesToCopy) {
    const src = path.join(__dirname, file);
    const dest = path.join(__dirname, 'lambda-temp', file);
    
    if (fs.statSync(src).isDirectory()) {
      fs.cpSync(src, dest, { recursive: true });
    } else {
      fs.copyFileSync(src, dest);
    }
  }

  // Install production dependencies
  console.log('📦 Installing production dependencies...');
  console.log('   (This may take a minute...)\n');
  execSync('npm install --production --prefix lambda-temp', {
    stdio: 'inherit'
  });

  // Create ZIP (platform-specific)
  console.log('\n📦 Creating lambda-package.zip...');
  
  if (process.platform === 'win32') {
    // Windows: Use PowerShell with absolute path
    const tempPath = path.join(__dirname, 'lambda-temp', '*');
    const zipPath = path.join(__dirname, 'lambda-package.zip');
    execSync(
      `powershell -command "Compress-Archive -Path '${tempPath}' -DestinationPath '${zipPath}' -Force"`,
      { stdio: 'inherit', cwd: __dirname }
    );
  } else {
    // Linux/Mac: Use zip
    execSync('cd lambda-temp && zip -r ../lambda-package.zip . -q', {
      stdio: 'inherit'
    });
  }

  // Check size
  const stats = fs.statSync('lambda-package.zip');
  const sizeMB = (stats.size / (1024 * 1024)).toFixed(2);
  
  console.log(`\n✅ Package created: lambda-package.zip (${sizeMB} MB)`);
  
  // Clean up temp directory
  console.log('🧹 Cleaning up temp files...');
  fs.rmSync('lambda-temp', { recursive: true, force: true });

  console.log('\n🎉 Done! Ready to upload to AWS Lambda.\n');
  console.log('📋 Next steps:');
  console.log('   1. Go to AWS Lambda console');
  console.log('   2. Upload lambda-package.zip');
  console.log('   3. Test your function\n');

} catch (error) {
  console.error('\n❌ Error:', error.message);
  process.exit(1);
}

File: package-lambda.py
#!/usr/bin/env python3

import zipfile
import os
import shutil
from pathlib import Path

print('\n📦 Packaging for AWS Lambda...\n')

# Clean up
print('🧹 Cleaning up previous package...')
if os.path.exists('lambda-package.zip'):
    os.remove('lambda-package.zip')
if os.path.exists('lambda-temp'):
    shutil.rmtree('lambda-temp')

# Create temp directory
print('📁 Creating package directory...')
os.makedirs('lambda-temp')

# Copy files
print('📄 Copying source files...')
files_to_copy = [
    ('lambda-handler.js', 'lambda-handler.js'),
    ('src', 'src'),
    ('package.json', 'package.json')
]

for src, dest in files_to_copy:
    src_path = Path(src)
    dest_path = Path('lambda-temp') / dest
    
    if src_path.is_dir():
        shutil.copytree(src_path, dest_path)
    else:
        shutil.copy2(src_path, dest_path)

# Install dependencies
print('📦 Installing production dependencies...')
print('   (This may take a minute...)\n')
os.system('npm install --production --prefix lambda-temp')

# Files to exclude (reduce size)
exclude_patterns = [
    '__pycache__',
    '.pytest_cache',
    '*.pyc',
    'test',
    'tests',
    'examples',
    'docs',
    '.git',
    '*.md',
    'LICENSE',
    'CHANGELOG',
]

def should_exclude(path):
    """Check if file should be excluded"""
    path_str = path.lower()
    for pattern in exclude_patterns:
        if pattern in path_str:
            return True
    return False

# Create ZIP with ZIP64 support (for large files)
print('\n📦 Creating lambda-package.zip...')
with zipfile.ZipFile('lambda-package.zip', 'w', zipfile.ZIP_DEFLATED, allowZip64=True) as zipf:
    file_count = 0
    for root, dirs, files in os.walk('lambda-temp'):
        for file in files:
            file_path = os.path.join(root, file)
            arcname = os.path.relpath(file_path, 'lambda-temp').replace('\\', '/')
            
            # Skip excluded files
            if not should_exclude(arcname):
                zipf.write(file_path, arcname)
                file_count += 1
    
    print(f'   Added {file_count} files')

# Get size
size_mb = os.path.getsize('lambda-package.zip') / (1024 * 1024)

print(f'Package created: lambda-package.zip ({size_mb:.2f} MB)')

# Clean up
print('Cleaning up temp files...')
shutil.rmtree('lambda-temp')

print('\nDone! Ready to upload to AWS Lambda.\n')
print('Next steps:')
print('   1. Go to AWS Lambda console')
print('   2. Upload lambda-package.zip')
print('   3. Test your function\n')

